SPEC.md

```diff
@@ -153,6 +154,47 @@ The network configuration is described in JSON form. The configuration can be st
   - `options` (list of strings): list of options that can be passed to the resolver
 
 Plugins may define additional fields that they accept and may generate an error if called with unknown fields. The exception to this is the `args` field may be used to pass arbitrary data which may be ignored by plugins.
+
+### Plugin Chaining and Delegation
+
+Multiple plugins may be chained together to perform more complex network configuration. IPAM plugins already operate this way, with the outer plugin delegating to the IPAM plugin and receiving its IPAM result.
+The network configuration may contain an `inner` field, which itself is a complete CNI network configuration block (except for the `cniVersion` and `name` fields, which cannot be specified in the `inner` dictionary).
+When the outer plugin finds the `inner` field, it proceeds to execute the inner plugin, passing the entire `inner` configuration via stdin, merging the top-level `cniVersion` and `name` fields.
+The inner plugin then executes, including possibly running yet another inner plugin if the top-level `inner` dictionary contains a second `inner` dictionary, specifying a third plugin.
+Each plugin then receives the IPAM result of the called plugin, and can use that result and optionally manipulate it before returning the result to its caller.
+
+For example, suppose traffic shaping should be enabled for a container whose network is configured with the standard `bridge` plugin and `host-local` IPAM plugins.
+
+```json
+{
+  "cniVersion": "0.2.0",
+  "name": "chaining-example",
+  "type": "traffic-shaping",
+  "ingressbw": "20M",
+  "egressbw": "10M",
```

Comment:
Say if there is a traffic-shaping plugin, does it make more sense to take ingress/egress bandwidth configs thru runtime configs instead of network configs? In k8s CNI network plugin, we do not manipulate net config once it is settled. How can we handle configs like this? What is the boundary between runtime configs and network configs?